azkaban-exec-server/src/main/java/azkaban/execapp/ProjectCacheCleaner.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2019 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.execapp;
+
+import azkaban.utils.FileIOUtils;
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.FilenameFilter;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```

Comment:
Another option is to have cleaning done in a separate thread. It could check if the usage has exceeded some threshold (perhaps 85%), and start removing project directories if so. It could remove directories until 50% of the space is free.

Also, it should be possible to safely remove older versions for projects -- only the latest one is saved in Azkaban, and executed for newer executions.